src/persist/src/indexed/trace.rs

```diff
@@ -310,9 +310,8 @@ impl Snapshot<Vec<u8>, Vec<u8>> for TraceSnapshot {
     fn read<E: Extend<((Vec<u8>, Vec<u8>), u64, isize)>>(&mut self, buf: &mut E) -> bool {
         if let Some(batch) = self.updates.pop() {
             buf.extend(batch.updates.iter().cloned());
-            return true;
         }
-        false
+        !self.updates.is_empty()
```

Comment:
I think the semantics of "read returns true if data was read in OR more data will be read in, read returns false if no data was read in AND no further data will be available" make sense / would work. I kept the semantics as currently written because I felt like that was a smaller change overall.

Reply:
sorry let me be more specific - it was conceptually simpler for me to just stick to the already defined and agreed upon semantics rather than trying to figure out a new semantics and a new impl of those for BufferSnapshot which would i think have to change its state to remember the first time someone called read on it. Neither of those things are the end of the world but its nice to have a already crisp definition of what I'm trying to build when I'm also trying to sort out a bunch of other bugs :)

do you think its ok to address the snapshot semantics change in a followup pr?

Reply:
Ha! Yeah I have a slightly different take (which is wildly less defensible than yours) where the behavior of trace/future/indexed/decoded snapshot was the contract and the rustdoc on Snapshot::read and the impl of BufferSnapshot were the ones that were wrong. I'm fine leaving this as a followup, but toss a TODO in there please.

Btw, I think a workable impl for BufferSnapshot is
```rust
impl Snapshot<Vec<u8>, Vec<u8>> for BufferSnapshot {
    fn read<E: Extend<((Vec<u8>, Vec<u8>), u64, isize)>>(&mut self, buf: &mut E) -> bool {
        let had_data = self.1.len() > 0;
        buf.extend(self.1.drain(..));
        had_data
    }
}
```

Reply:
Added the todo!